frontend/exporter/src/types/ty.rs

```diff
@@ -1272,21 +1272,57 @@ pub enum AliasRelationDirection {
 }
 
 /// Reflects [`ty::ClosureArgs`]
-#[derive(AdtInto)]
-#[args(<'tcx, S: UnderOwnerState<'tcx> >, from: ty::ClosureArgs<ty::TyCtxt<'tcx>>, state: S as s)]
-#[derive(Clone, Debug, JsonSchema)]
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, JsonSchema)]
 #[derive_group(Serializers)]
 pub struct ClosureArgs {
-    #[value(self.kind().sinto(s))]
+    /// The base kind of this closure. The kinds are ordered by inclusion: any `Fn` works as an
+    /// `FnMut`, and any `FnMut` works as an `FnOnce`.
     pub kind: ClosureKind,
-    #[value(self.parent_args().sinto(s))]
+    /// The arguments to the parent (i.e., the item which defines the closure).
     pub parent_args: Vec<GenericArg>,
-    #[value(self.sig().sinto(s))]
-    pub sig: PolyFnSig,
-    #[value(self.upvar_tys().sinto(s))]
+    /// The solved predicates from the parent (i.e., the item which defines the closure).
+    pub parent_trait_refs: Vec<ImplExpr>,
+    /// The proper `fn(A, B, C) -> D` signature of the closure.
+    pub untupled_sig: PolyFnSig,
+    /// The signature of the closure as one input and one output, where the input arguments are
+    /// tupled. This is relevant to implementing the `Fn*` traits.
+    pub tupled_sig: PolyFnSig,
+    /// The set of captured variables. Together they form the state of the closure.
     pub upvar_tys: Vec<Ty>,
 }
 
+#[cfg(feature = "rustc")]
+impl ClosureArgs {
+    // Manual implementation because we need the `def_id` of the closure.
+    pub(crate) fn sfrom<'tcx, S>(
+        s: &S,
+        def_id: RDefId,
+        from: ty::ClosureArgs<ty::TyCtxt<'tcx>>,
+    ) -> Self
+    where
+        S: UnderOwnerState<'tcx>,
+    {
+        let tcx = s.base().tcx;
+        let sig = from.sig();
+        ClosureArgs {
+            kind: from.kind().sinto(s),
+            parent_args: from.parent_args().sinto(s),
+            // The solved predicates from the parent (i.e., the item which defines the closure).
+            parent_trait_refs: {
+                // TODO: handle nested closures
```

Comment:
Shall we file a follow up issue?

Reply:
https://github.com/hacspec/hax/issues/1217